src/point.rs

```diff
@@ -101,12 +101,10 @@ impl<T: Zero, U> Point2D<T, U> {
     pub fn zero() -> Self {
         Self::origin()
     }
-}
 
-impl<T: Copy + Zero, U> Point2D<T, U> {
-    /// Convert into a 3d point.
+    /// Convert into a 3d point with z-coordinate equals to zero.
     #[inline]
-    pub fn to_3d(&self) -> Point3D<T, U> {
+    pub fn to_3d(self) -> Point3D<T, U> {
```

Comment:
I don't see what purpose of this change. It comes as a surprise to me that `to_*` conversion methods should consume self (that's why `into_` methods exist). There are similar conversion methods that take self by reference in the standard library, for example [`Cstr::to_str`](https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.to_str).

It seems to me that while `into_` methods are conventionally consuming self, `to_` ones don't necessarily. Whether they take self by value or reference depends on how cheap the receiver type is to pass by value.

I don't really mind in the context of points vectors and sizes, since we already pass them by value arguments in most places, but we decided to pass larger structures like rectangles and matrices by reference and that won't change.
If we are to do this, let's be consistent do it for all `Point2D` methods.

Reply:
> but we decided to pass larger structures like rectangles and matrices by reference and that won't change.

Actually, they already consume `self`, because all changed methods use all fields of struct to create another struct. Because of `Copy` bounds you actually makes copy of whole struct. So that is logical step -- remove `Copy` bound and consume object. For `Copy` objects nothing changes, but with such change method will also work for non-Copy types.

For this reason, I have changed only those methods that actually completely consume the old object.

Reply:
It's mostly a question of calling conventions and the overhead of calling a function (that aren't inlined). Large parameters when passed by values are copied on the stack and the function receives an address while small parameters tend to be passed via registers.

Reply:
I think, that non-copyable types always moved, and for copyable compiler also can use move if it can prove that copy is unnecessary. At least [this comparison](https://rust.godbolt.org/z/oaYszh) do not show much difference (although I am not a specialist in making comparison and analyzing compiler output).

> There are similar conversion methods that take self by reference in the standard library, for example Cstr::to_str.

IMO this is not an equivalent comparison because this method returns a reference to the internals of `CStr`. More relevant example will be [Wrapping::to_be](https://doc.rust-lang.org/std/num/struct.Wrapping.html#method.to_be) and similar [`to_...`](https://doc.rust-lang.org/std/?search=to_) methods. Of course, it would be better name these methods `into_...`, but that will be strict breaking change...

> If we are to do this, let's be consistent do it for all Point2D methods.

Only the `lerp` and `add_size` methods remained. I plan to remove `add_size` when doing refactoring of `rect.rs` (which first is require refactoring of `point.rs` and `size.rs`), and `lerp` still requires at least a `Clone` bound therefore consume has no sense.

I didn't quite understand your position on this change (perhaps just my understanding of English isn't enough): do you mind replacing `&self` with `self`?

Reply:
> More relevant example will be Wrapping::to_be and similar to_... methods. 

That's what I was saying: these methods are on very basic types and like all methods on ints, floats and such, they take self by value (it's not because they are conversion methods).

My position is:
I don't think we need to change from taking `&self` to taking `self`, in the sense that as a user of euclid it doesn't have any benefit. I disagree that it is more idiomatic for `to_*` methods to take self by value. 
The only two reasons to chose between `self` and `&self` are:
 - whether it is efficient to consume the object, for example to avoid reallocating memory.
 - whether it is efficient to call the function which depends on the size of the struct.

Since points, vectors and sizes are rather small objects, it's not particularly expensive to pass them by value even if the method is not inlined (at least for 32 bit floats and ints, though it will probably be more expensive with `f64`), that's why various methods in euclid already pass them as value arguments, I am not opposed to also pass `self` as value for the same reason (after all for the compiler self is just a function argument).

So if you really want to change point methods to take `self`, then OK, however:
 - All methods on points, vectors and sizes have to take `self` if we do this.
 - This is a breaking change so it will have to wait for other motivating breaking changes to be worth the hassle of updating downstream dependencies. There is enough code depending on euclid that even seemingly innocuous breaking changes are likely to cause some code to stop compiling.
 - methods on bigger objects like rectangles and matrices won't take self by value. This is really not a question of conversion methods but a question of whether it is efficient to take self by value or reference for the size of an object.

Reply:
> This is a breaking change so it will have to wait for other motivating breaking changes to be worth the hassle of updating downstream dependencies.

Of course. I'm not suggesting we do this in release 0.20.x. The only thing I don't understand is why can't you release 0.21 straight away? This does not mean that everyone should immediately update to it.

> methods on bigger objects like rectangles and matrices won't take self by value. This is really not a question of conversion methods but a question of whether it is efficient to take self by value or reference for the size of an object.

As far as I know, receiving an object by value does not mean that the compiler is prohibited from using reference, especially for `self`, because using a reference is how a virtual function call is implemented. In Rust, passing by value means only a hint for borrow checker. [As you can see](https://rust.godbolt.org/z/J-KhTX), assembler code in the `main` is the same.

Therefore, there is no reason not to use pass-by-value if more generic code can be achieved. This is what Rust holds on to

Reply:
> If we are to do this, let's be consistent do it for all Point2D methods.

I removed an `add_size` method, because if change it arguments to accept objects by-value, it will be similar to ordinal `Add` operator. Also I change `lerp` signature and relaxed bound from `Copy` to `Clone` to avoid making extra copies when they are not needed. I think the code from this hasn't become much more complicated and is still very easy to understand.

So there are no more methods left in `Point` that accept point by reference.

Reply:
> As far as I know, receiving an object by value does not mean that the compiler is prohibited from using reference, especially for self, because using a reference is how a virtual function call is implemented. In Rust, 

> As you can see, assembler code in the main is the same.

What you are showing me is a very short function which is inlined by the compiler.

> passing by value means only a hint for borrow checker. 

This is not true. If the function is not inlined, passing by value or by reference are two different things defined by the calling convention for the target platform. The compiler has to stick to the calling convention otherwise things like shared shared libraries wouldn't be able to inter-operate. When the function is inlined, well there is by definition no parameter passing at all.

> Of course. I'm not suggesting we do this in release 0.20.x. The only thing I don't understand is why can't you release 0.21 straight away? This does not mean that everyone should immediately update to it.

We have a fairly big ecosystem of crates within Servo and Firefox that we keep in sync and managing multiple branches (on a crate that is imported by most others) makes that complicated. Also I don't even know how that would work with the bors-servo workflow for integrating pull requests.
In practice, bumping euclid always turns into a game of domino and everyone else soon end up having to update. So we do our best to only publish breaking changes when the value we get out of it compensates the troubles.
I am not saying that we have to wait forever, there are a few breaking changes I plan to make that I have been postponing for the same reasons.

I realize how frustrating this is when proposing changes that are harmless and even beneficial as many of the things you are contributing here.

If you can extract the non-breaking parts of these pull requests, we can land and publish them soon. Then we take the time to work out the rest.

Reply:
>> [As you can see](https://rust.godbolt.org/z/J-KhTX), assembler code in the main is the same.
>
> What you are showing me is a very short function which is inlined by the compiler.

In exactly that case I specially use `#[inline(never)]` and compiler truly inserts function call. And I use relatively big object, which is bigger, than `Transform3D`, the most big object in euclid. Since the compiler does not inline the function, there is no meaning to what size it has.

> In practice, bumping euclid always turns into a game of domino and everyone else soon end up having to update.

Well, it's sad to hear that. I hope the situation will improve in the future.

> If you can extract the non-breaking parts of these pull requests, we can land and publish them soon. Then we take the time to work out the rest.

Okay, I'll do that. I'll leave my PRs open and when you decide to make breaking changes on time, just comment on them and I'll update my PRs if required.

Reply:
> In exactly that case I specially use #[inline(never)] and compiler truly inserts function call. And I use relatively big object, which is bigger, than Transform3D, the most big object in euclid. Since the compiler does not inline the function, there is no meaning to what size it has.

Your godbolt snippet actually agrees with me. It is easier to see if instead of calling `by_value(big)` once, you call it three times:

```rust
by_value(big);
by_value(big);
by_value(big);
```

If you look at the assembly, you will see that the data for that struct is pushed on the stack before each function call:

```asm
        movaps  xmmword ptr [rsp + 112], xmm0
        movaps  xmmword ptr [rsp + 96], xmm0
        movaps  xmmword ptr [rsp + 80], xmm0
        movaps  xmmword ptr [rsp + 64], xmm0
        movaps  xmmword ptr [rsp + 48], xmm0
        movaps  xmmword ptr [rsp + 32], xmm0
        movaps  xmmword ptr [rsp + 16], xmm0
        movaps  xmmword ptr [rsp], xmm0
        mov     rdi, rsp
        call    example::by_value
        xorps   xmm0, xmm0
        movaps  xmmword ptr [rsp + 112], xmm0
        movaps  xmmword ptr [rsp + 96], xmm0
        movaps  xmmword ptr [rsp + 80], xmm0
        movaps  xmmword ptr [rsp + 64], xmm0
        movaps  xmmword ptr [rsp + 48], xmm0
        movaps  xmmword ptr [rsp + 32], xmm0
        movaps  xmmword ptr [rsp + 16], xmm0
        movaps  xmmword ptr [rsp], xmm0
        mov     rdi, rsp
        call    example::by_value
        xorps   xmm0, xmm0
        movaps  xmmword ptr [rsp + 112], xmm0
        movaps  xmmword ptr [rsp + 96], xmm0
        movaps  xmmword ptr [rsp + 80], xmm0
        movaps  xmmword ptr [rsp + 64], xmm0
        movaps  xmmword ptr [rsp + 48], xmm0
        movaps  xmmword ptr [rsp + 32], xmm0
        movaps  xmmword ptr [rsp + 16], xmm0
        movaps  xmmword ptr [rsp], xmm0
        mov     rdi, rsp
        call    example::by_value
```

Then replace these three by_value calls with three by_ref(&big) calls

```rust
by_ref(big);
by_ref(big);
by_ref(big);
```

And it simply moves the address of the struct into a register before calling the function each time:

```asm
        mov     rdi, rbx
        call    example::by_ref
        mov     rdi, rbx
        call    example::by_ref
        mov     rdi, rbx
        call    example::by_ref
```

> Okay, I'll do that. I'll leave my PRs open and when you decide to make breaking changes on time, just comment on them and I'll update my PRs if required.

Thanks a lot!